vispy/visuals/_scalable_textures.py

```diff
@@ -0,0 +1,441 @@
+
+    This class includes optimizations where image data is not re-normalized
+    if the previous normalization can still be used to visualize the data
+    with the new color limits.
+
+    This class should only be used internally. For similar features where
+    scaling occurs on the GPU see
+    :class:`vispy.visuals._scalable_textures.GPUScaledTextureMixin`.
+
+    To use this mixin, a subclass should be created to combine this mixin with
+    the texture class being used. Existing subclasses already exist in this
+    module. Note that this class **must** appear first in the subclass's parent
+    classes so that its ``__init__`` method is called instead of the parent
+    Texture class.
+
+    """
+
+    def __init__(self, data=None, **texture_kwargs):
+        self._data_limits = None
+        # Call the __init__ of the mixin base class
+        super().__init__(data, **texture_kwargs)
+
+    def _clim_outside_data_limits(self, cmin, cmax):
+        if self._data_limits is None:
+            return False
+        return cmin < self._data_limits[0] or cmax > self._data_limits[1]
+
+    def set_clim(self, clim):
+        """Set clim and return if a texture update is needed."""
+        need_texture_upload = False
+        # NOTE: Color limits are not checked against data type limits
+        if isinstance(clim, str):
+            if clim != 'auto':
+                raise ValueError('clim must be "auto" if a string')
+            need_texture_upload = True
+            self._clim = clim
+        else:
+            try:
+                cmin, cmax = clim
+            except (ValueError, TypeError):
+                raise ValueError('clim must have two elements')
+            if self._clim_outside_data_limits(cmin, cmax):
+                need_texture_upload = True
+            self._clim = (cmin, cmax)
+        return need_texture_upload
+
+    @property
+    def clim_normalized(self):
+        """Normalize current clims to match texture data inside the shader.
+
+        If data is scaled on the CPU then the texture data will be in the range
+        0-1 in the _build_texture() method. Inside the fragment shader the
+        final contrast adjustment will be applied based on this normalized
+        ``clim``.
+
+        """
+        range_min, range_max = self._data_limits
+        clim_min, clim_max = self.clim
+        clim_min = (clim_min - range_min) / (range_max - range_min)
+        clim_max = (clim_max - range_min) / (range_max - range_min)
+        return clim_min, clim_max
+
+    @staticmethod
+    def _scale_data_on_cpu(data, clim, copy=True):
+        if copy:
+            should_cast_to_f32(data.dtype)
+            data = np.array(data, dtype=np.float32, copy=copy)
+        elif not copy and not np.issubdtype(data.dtype, np.floating):
+            raise ValueError("Data must be of floating type for no copying to occur.")
+
+        if clim[0] == clim[1]:
+            if clim[0] != 0:
+                data /= clim[0]
```

Comment:
What is this for? :)

Reply:
I honestly don't know. It was in the original code and I think is to make it so if the limits are the same you don't end up with a divide by zero. I guess it "normalizes" the data to about the same range as the user expected with their color limits being set to the same thing? I didn't change it because I assumed that someone might be depending on it. I'm definitely open to changing it and it doesn't make a lot of sense to me.

Reply:
haha, ok lets not touch it for now :)

Reply:
see https://github.com/vispy/vispy/commit/8d280f2b50c0fce92a1bc35e5229f223293eadba for context ;-)

Reply:
Thanks! That makes sense. Let's add a comment while we're here.

Reply:
I still don't get it. Like I get what it is talking about. If clims are the same then we get a divide by zero at some point. If clims are the same what should the expected output be? This line `vol *= 1.0 / self._clim[0]` from the original code (same as my `data /= clim[0]` above) just doesn't make sense as far as logical output in my opinion. I guess why divide at all?

Reply:
> I guess why divide at all?

Good point. The division is ... arbitrary.

Reply:
We could do something like: if clims are the same, make data all 0s (`np.zeros_like`) and make clims 0 and 1. Or...we could detect the clims being the same in the shader and discard the fragment. That should make it work for CPU or GPU scaled textures. I'd just have to make sure that `.clim_normalized` returns clims that are equal to each other. Thoughts?

Reply:
If the clims are equal, we'd expect the result to be all gray, I guess. We should not have to change the data. So maybe something like:
```py
if clim[0] == clim[1]:
  clim = clim[0] - 0.5, clim[0] + 0.5
```

Another thought: let's do this in a separate PR :P

Reply:
oh wait, this code is in the CPUScaledTexture. So I guess setting the data to all 127 makes sense, and something like what I wrote above for the GPUScaledTexture.

Reply:
> If the clims are equal, we'd expect the result to be all gray, I guess.

Eh. That's like saying `x / 0` is a real number. I'm starting to think the two texture subclasses shouldn't do any scaling if the clims are the same and leave it to the shader code to decide what to do. I think for both the ImageVisual and the VolumeVisual it should do one of the following:

1. discard (nothing there, fully transparent)
2. return 0 so the colormap will get its lowest value
3. return 0.5 so the colormap will get its middle value

Thoughts? @jni @tlambert03 @alisterburt Any thoughts on this? What do you expect as output for the Image or Volume if the color limits are equal to each other? Does that result change if the limits remain equal to each other but change in value (ex. `(0, 0)` versus `(0.5, 0.5)` versus `(1.0, 1.0)`)?

Reply:
> Eh. That's like saying x / 0 is a real number

I think what I meant to say is what you call option nr 3 :)

Reply:
In matplotlib with no vmin/vmax on imshow:

![image](https://user-images.githubusercontent.com/1828519/121443137-8de49080-c952-11eb-9928-e59c403e9220.png)

With vmin/vmax both set to 0, or both set to 255, or both set to 128, you get this for each:

![image](https://user-images.githubusercontent.com/1828519/121443525-427eb200-c953-11eb-883b-3f392bdc7864.png)

So matplotlib always shows the lowest value in the colormap.

Reply:
So based on my twitter poll (18 responses) most people think this should raise an error, while the second most popular option follows mpl's behavior above. I'm starting to think I agree with mpl. This makes sense in the case the limits are automatically detected and all input pixels are the same. You'd want to see *something* on the screen even if it doesn't mean anything. I think in vispy the easiest way to do this is to not scale the data, detect the equal clims in the shader, and return 0 (to be colormapped).